File: 프로그래머스/unrated/169199. 리코쳇 로봇/리코쳇 로봇.py
from collections import deque
dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]

def solution(board):
    
    answer = 0
    cnt = 0
    visited = []
    for _ in range(len(board)):
        visited.append([0]*len(board[0]))
        
    def BFS(x, y):
        q = deque()
        q.append((x,y))
        visited[y][x]=1
        while q:
            x, y = q.popleft()
            
            if board[y][x]=="G":
                return visited[y][x]-1
            
            # 미끄러지기
            for k in range(4):
                # 처음 값 기억하기
                nx, ny = x, y
                while True:
                    nx = nx + dx[k]
                    ny = ny + dy[k]
                    # 처음 방문하는 장애물일 경우
                    if 0<=nx<len(board[0]) and 0<=ny<len(board) and board[ny][nx]== "D":
                        nx -= dx[k]
                        ny -= dy[k]
                        break
                    # 처음 방문하는 배열 경계일 경우
                    if nx<0 or nx>=len(board[0]) or ny<0 or ny>=len(board):
                        nx -= dx[k]
                        ny -= dy[k]
                        break
                if not visited[ny][nx]:
                    visited[ny][nx] = visited[y][x] + 1
                    q.append((nx,ny))
        return -1
    
    for i in range(len(board)):
        for j in range(len(board[0])):
            if board[i][j] == "R":
                answer = BFS(j,i)
    
    return answer

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.js
function solution(board) {
    const dir = [[0,1],[0,-1],[1,0],[-1,0]]
    var answer = 0;
    const row = board.length
    const col = board[0].length
    const start = [0,0]
    const end = [0,0]
    const visited = Array.from({length : row}, () => Array.from({length : col}, ()=> false))
    
    for(let i = 0; i < row; i++){
        for(let j = 0; j < col; j++){
            if(board[i][j]==="R"){
                start[0] = i
                start[1] = j
            }
            if(board[i][j]==="G"){
                end[0] = i
                end[1] = j
            }
        }
    }
        
    const queue = []
    queue.push([start[0], start[1], 0])
    visited[start[0]][start[1]] = true
    let begin = 0
    
    while(begin < queue.length){
        const [cy, cx, cnt] = queue[begin++]
        
        if(cy === end[0] && cx === end[1]) return cnt
        
       
        for(let k = 0; k < 4; k++){
             let ny = cy 
             let nx = cx 
             while(true) {
                const ty = ny + dir[k][0];
                const tx = nx + dir[k][1];
                
                if(ty < 0 || ty >= row || tx < 0 || tx >= col || board[ty][tx] === "D" ) break;
                ny = ty;
                nx = tx;
            }
            
            if(visited[ny][nx]) continue
            visited[ny][nx] = true
            queue.push([ny, nx, cnt + 1])
        }
    }
    
    return -1;
}